### PR TITLE
HA service/lease refactor

### PR DESCRIPTION
**Pull request by antekresic**
Service contained a lot of business logic which should be inside
of lease. This refactor moves it to the proper place and simplifies
the code by removing the additional complexity caused by the logic
misplacement.

### CODE REVIEW

File: pkg/ha/state/lease.go [hunk 1]
```diff
@@ -69,44 +75,117 @@ func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT,
...
+
+	if err := l.TryChangeLeader(currT); err != nil {
+		return false, time.Time{}, err
+	}
+	return l.confirmLeaseValidation(replica, minT)
+}
+
+func (l *Lease) confirmLeaseValidation(replica string, minT time.Time) (bool, time.Time, error) {
+	l._mu.RLock()
+	defer l._mu.RUnlock()
```
Comment (cevian) [inline pkg/ha/state/lease.go:105]: Nit: I like following this convention throughout this code
```suggestion
	l._mu.RLock()
	leader := l.leader
	leaseStart := l.leaseStart
	l._mu.RUnlock()
```

File: pkg/ha/state/lease.go [hunk 2]
```diff
@@ -69,44 +75,117 @@ func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT,
 	}, nil
 }
 
-// UpdateLease uses the supplied client to attempt to update the lease for the potentialLeader.
+func (l *Lease) ValidateSamplesInfo(replica string, minT, maxT, currT time.Time) (bool, time.Time, error) {
+	l._mu.RLock()
+	currentLeader := l.leader
+	currentDeadline := l.leaseUntil
```
Comment (cevian) [inline pkg/ha/state/lease.go:81]: nit: keep names consistentt
```suggestion
	leaseUntil := l.leaseUntil
```

File: pkg/ha/state/lease.go [hunk 3]
```diff
@@ -69,44 +75,117 @@ func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT,
...
-	cluster := h.cluster
-	maxTimeInstance := h.maxTimeInstance
-	maxTimeSeen := h.maxTimeSeen
-	h._mu.RUnlock()
-	leaseState, err := c.TryChangeLeader(
-		context.Background(), cluster, maxTimeInstance, maxTimeSeen,
+func (l *Lease) TryChangeLeader(currT time.Time) error {
+	l._mu.RLock()
+	cluster := l.cluster
+	currentDeadline := l.leaseUntil
```
Comment (cevian) [inline pkg/ha/state/lease.go:153]: again:
```suggestion
	leaseUntil := l.leaseUntil
```

File: pkg/ha/state/lease.go [hunk 4]
```diff
@@ -69,44 +75,117 @@ func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT,
...
+	maxTimeInstance := l.maxTimeInstance
+	maxTimeSeen := l.maxTimeSeen
+	recentLeaderSeen := l.recentLeaderWriteTime
+	l._mu.RUnlock()
+
```
Comment (cevian) [inline pkg/ha/state/lease.go:158]:
```suggestion
	// only proceed if we have seen samples after the lease expired
	// and haven't seen samples from leader in haLastWriteInterval (wall-clock-time)	
```

File: pkg/ha/state/lease.go [hunk 5]
```diff
@@ -69,44 +75,117 @@ func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT,
...
-		return nil, fmt.Errorf("could not call try change leader from db: %#v", err)
+		return fmt.Errorf("could not call try change leader from db: %#v", err)
 	}
 
-	return h.setUpdateFromDB(leaseState), nil
+	return nil
+}
+
+func (l *Lease) changeLeader(cluster string, leader string, maxTimeSeen time.Time) error {
+	leaseState, err := l.client.TryChangeLeader(
```
Comment (cevian) [inline pkg/ha/state/lease.go:179]:
```suggestion
	stateFromDB, err := l.client.TryChangeLeader(
```

Comment (cevian) [inline pkg/ha/state/lease.go:179]: should l.client be protected by Mutex? If not, can we make it clear what should and shouldn't be protected by mutex (maybe by putting the mutex above those fields it protects in the struct definition)?

File: pkg/tests/end_to_end_tests/ha_single_promscale_test.go [hunk 6]
```diff
@@ -101,7 +105,9 @@ func prepareIngestorWithHa(db *pgxpool.Pool, t testing.TB) (*util.ManualTicker,
 	// so the calls to ha.Service.checkLeaseUntilAndLastWrite
 	// depends only on the max time seen by an instance
 	// and try_change_leader is called when we want
-	tooFarInTheFutureNowFn := func() time.Time { return time.Now().Add(time.Hour) }
+	tooFarInTheFutureNowFn := func() time.Time {
+		return time.Now().Add(time.Duration(atomic.AddInt32(&tCount, 1)) * time.Hour)
```
Comment (cevian) [inline pkg/tests/end_to_end_tests/ha_single_promscale_test.go:109]: why was this change needed?

File: pkg/ha/state/lease.go [hunk 7]
```diff
@@ -69,44 +75,117 @@ func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT,
...
+	maxTimeInstance := l.maxTimeInstance
+	maxTimeSeen := l.maxTimeSeen
+	recentLeaderSeen := l.recentLeaderWriteTime
+	l._mu.RUnlock()
+
+	if currentDeadline.After(maxTimeSeen) {
+		return nil
+	}
+
+	if currT.Sub(recentLeaderSeen) <= haLastWriteInterval {
```
Comment (cevian) [inline pkg/ha/state/lease.go:163]: can we add a test that would catch the bug we had here before?

**Review by cevian — CHANGES_REQUESTED** [review]
Overall looks very good. Some minor nits and a few questions.

File: pkg/tests/end_to_end_tests/ha_single_promscale_test.go (hunk 6 above)
Comment (antekresic) [reply pkg/tests/end_to_end_tests/ha_single_promscale_test.go:109]: This is from a bug fix PR, will be removed from this commit once that PR is merged in (#569 )

File: pkg/ha/state/lease.go (hunk 7 above)
Comment (antekresic) [reply pkg/ha/state/lease.go:163]: Sure, in a separate commit.

Comment (VineethReddy02) [reply pkg/ha/state/lease.go:163]: @antekresic  can I know more info on the bug we noticed here? I am curious to know how did I not find it in my manual tests...

Comment (VineethReddy02) [reply pkg/ha/state/lease.go:163]: Ahh, just noticed it here https://github.com/timescale/promscale/pull/569

File: pkg/pgmodel/ingestor/parser.go [hunk 8]
```diff
@@ -5,27 +5,42 @@
...
 	"github.com/timescale/promscale/pkg/pgmodel/common/errors"
 	"github.com/timescale/promscale/pkg/pgmodel/model"
 	"github.com/timescale/promscale/pkg/prompb"
 )
 
-type Parser interface {
-	ParseData([]prompb.TimeSeries) (map[string][]model.Samples, int, error)
+// Data parser is responsible for parsing incoming request data into the
+// appropriate format and filter it using the appropriate filter.
+type DataParser struct {
```
Comment (cevian) [inline pkg/pgmodel/ingestor/parser.go:17]: optional: Can we just call this Parser?

File: pkg/ha/ha_parser.go [hunk 9]
```diff
@@ -6,69 +6,67 @@ package ha
```
Comment (cevian) [inline pkg/ha/ha_parser.go:6]: this file should be renamed filter.go

File: pkg/ha/ha_service.go [hunk 10]
```diff
@@ -17,26 +17,10 @@ import (
```
Comment (cevian) [inline pkg/ha/ha_service.go:17]: can we rename ha_service.go->service.go

**Review by cevian — APPROVED** [review]
Approved with some needed changes. Especially removal of the diff file.

File: pkg/ha/filter.go [hunk 11]
```diff
@@ -6,69 +6,67 @@ package ha
...
+// gets from the lease service.
+type Filter struct {
 	service *Service
-	scache  cache.SeriesCache
 }
 
-func NewHAParser(service *Service, scache cache.SeriesCache) *haParser {
-	return &haParser{
+// NewFilter creates a new Filter based on the provided Service.
+func NewFilter(service *Service) *Filter {
```
Comment (atanasovskib) [inline pkg/ha/filter.go:25]: 🔥
